### PR TITLE
UCS/SYS: Add ucs_sockaddr_cmp needed for TCP

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

1. Adds `ucs_sockaddr_cmp`
2. `ucs_sockaddr_is_equal` uses `ucs_sockaddr_cmp` under the hood
3. Refactor tests for is_equal and cmp routines

## Why ?

`ucs_sockaddr_cmp` will be used by TCP to compare socket addresses and determine how is lower to resolve simultaneous connection establishment situation.

## How ?

The `ucs_sockaddr_cmp` is almost the same as `ucs_sockaddr_is_equal` was. It compares in_addr/in6_addr using `memcmp()` and returns signum of the value; ports are compared using `<`.

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
this is backported from #3547 PR (#3547 will be re-based on this then)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10191/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10196/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:mlx:retest

File: src/ucs/sys/sock.c [hunk 1]
```diff
@@ -154,7 +154,8 @@ ucs_status_t ucs_socket_connect_nb_get_status(int fd)
     }
 
     if (conn_status != 0) {
-        ucs_error("SOL_SOCKET(SO_ERROR) status on fd %d: %s", fd, strerror(conn_status));
+        ucs_error("SOL_SOCKET(SO_ERROR) status on fd %d: %s",
```
Comment (brminich) [inline src/ucs/sys/sock.c]: plz, avoid unrelated changes next time

File: src/ucs/sys/sock.h [hunk 2]
```diff
@@ -262,12 +262,35 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+ * @param [in]     sa1        Pointer to sockaddr structure #1.
+ * @param [in]     sa2        Pointer to sockaddr structure #2.
+ * @param [in/out] status_p   Pointer (can be NULL) to a status: UCS_OK on success
+ *                            or UCS_ERR_INVALID_PARAM on failure.
+ *
+ * @return Returns an integral value indicating the relationship between the
+ *         socket addresses:
+ *         > 0 - the first socket address is greater than the second
+ *               socket address;
+ *         < 0 - the first socket address is greater than the second
```
Comment (brminich) [inline src/ucs/sys/sock.h]: copy paste from previous statement

File: src/ucs/sys/math.h [hunk 3]
```diff
@@ -85,6 +85,9 @@ BEGIN_C_DECLS
 #define ucs_div_round_up(_n, _d) \
     (((_n) + (_d) - 1) / (_d))
 
+#define ucs_signum(_n) \
```
Comment (brminich) [inline src/ucs/sys/math.h]: what is the purpose of this macro?

File: src/ucs/sys/sock.c [hunk 4]
```diff
@@ -354,31 +361,52 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
-        ucs_error("unknown address family: %d", sa1->sa_family);
-        status = UCS_ERR_INVALID_PARAM;
+        result = ucs_signum(memcmp(&UCS_SOCKET_INET6_ADDR(sa1),
+                                   &UCS_SOCKET_INET6_ADDR(sa2),
+                                   sizeof(UCS_SOCKET_INET6_ADDR(sa1))));
+
+        port1 = ntohs(UCS_SOCKET_INET6_PORT(sa1));
+        port2 = ntohs(UCS_SOCKET_INET6_PORT(sa2));
+
+        if (!result && (port1 != port2)) {
```
Comment (brminich) [inline src/ucs/sys/sock.c]: this code is the same in both cases, can be moved outside the switch

File: src/ucs/sys/sock.h [hunk 5]
```diff
@@ -262,12 +262,35 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+ *               socket address;
+ *         = 0 - the socket addresses are equal.
+ *         Note: it returns a postive integer value in case of error occured
```
Comment (brminich) [inline src/ucs/sys/sock.h]: positive

File: src/ucs/sys/sock.c [hunk 6]
```diff
@@ -354,31 +361,52 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+    if (!ucs_sockaddr_is_known_af(sa1) ||
+        !ucs_sockaddr_is_known_af(sa2)) {
+        ucs_error("unknown address family: %d",
+                  !ucs_sockaddr_is_known_af(sa1) ?
+                  sa1->sa_family : sa2->sa_family);
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    if (sa1->sa_family < sa2->sa_family) {
```
Comment (brminich) [inline src/ucs/sys/sock.c]: `<` -> `!=`

File: src/ucs/sys/math.h (hunk 3 above)
Comment (dmitrygladkov) [reply src/ucs/sys/math.h]: move to gtest - it is needed to compare expected and actual results

File: src/ucs/sys/sock.c (hunk 6 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: thx! fixed

File: src/ucs/sys/sock.c (hunk 4 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: done

File: src/ucs/sys/sock.h (hunk 2 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: yep, fixed

File: src/ucs/sys/sock.h (hunk 5 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10219/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10222/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10227/ for details (Mellanox internal link).

File: test/gtest/ucs/test_sock.cc [hunk 7]
```diff
@@ -278,21 +282,60 @@ static bool sockaddr_is_equal(int sa_family, const char *ip_addr1,
...
+    ASSERT_UCS_OK(status);
     status = ucs_sockaddr_set_port(sa2, port2);
-    EXPECT_UCS_OK(status);
+    ASSERT_UCS_OK(status);
 
-    result1 = ucs_sockaddr_is_equal(sa1, sa2, &status);
-    EXPECT_UCS_OK(status);
+    // `sa1` vs `sa2`
+    {
+        int addr_cmp_res = strcmp(ip_addr1, ip_addr2);
```
Comment (brminich) [inline test/gtest/ucs/test_sock.cc]: maybe better compare `UCS_SOCKET_INET_ADDR` with `memcmp`, like the target function does? Otherwise you are testing slightly different stuff than the target func. Therefore, in theory, the test can fail with no real error (depending on the ip).

Comment (dmitrygladkov) [reply test/gtest/ucs/test_sock.cc]: fixed

File: test/gtest/ucs/test_sock.cc [hunk 8]
```diff
@@ -11,6 +11,9 @@ extern "C" {
 
 #include <sys/un.h>
 
+#define signum(_n) \
```
Comment (brminich) [inline test/gtest/ucs/test_sock.cc]: you do not need it anymore

Comment (dmitrygladkov) [reply test/gtest/ucs/test_sock.cc]: removed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10231/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10232/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe could you review pls?

File: src/ucs/sys/sock.c [hunk 9]
```diff
@@ -354,37 +361,59 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+    }
+
 out:
     if (status_p) {
         *status_p = status;
     }
     return result;
 }
+
+int ucs_sockaddr_is_equal(const struct sockaddr *sa1,
```
Comment (yosefe) [inline src/ucs/sys/sock.c]: maybe we can remove this function now? it's just a wrapper  for ucs_sockaddr_cmp

File: src/ucs/sys/sock.c [hunk 10]
```diff
@@ -354,37 +361,59 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+        !ucs_sockaddr_is_known_af(sa2)) {
+        ucs_error("unknown address family: %d",
+                  !ucs_sockaddr_is_known_af(sa1) ?
+                  sa1->sa_family : sa2->sa_family);
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
 
     if (sa1->sa_family != sa2->sa_family) {
+        result = (sa1->sa_family < sa2->sa_family) ? -1 : 1;
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: good catch! done

File: src/ucs/sys/sock.c [hunk 11]
```diff
@@ -354,37 +361,59 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
-        status = UCS_ERR_INVALID_PARAM;
+        result = memcmp(&UCS_SOCKET_INET6_ADDR(sa1),
+                        &UCS_SOCKET_INET6_ADDR(sa2),
+                        sizeof(UCS_SOCKET_INET6_ADDR(sa1)));
+        port1 = ntohs(UCS_SOCKET_INET6_PORT(sa1));
+        port2 = ntohs(UCS_SOCKET_INET6_PORT(sa2));
         break;
     }
 
+    if (!result && (port1 != port2)) {
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: good catch! done

File: src/ucs/sys/sock.c (hunk 9 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: yes, we can. removed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10283/ for details (Mellanox internal link).
